File: backend/app/routers/shifts.py
from datetime import date
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from app.database import get_db
from app.dependencies import require_admin
from app.schemas.shift import ShiftCreate, ShiftUpdate, ShiftOut
from app.services import shift_service

router = APIRouter(prefix="/shifts", tags=["shifts"])


@router.get("", response_model=list[ShiftOut])
def list_shifts(
    employee_id: Optional[int] = Query(None),
    fecha_inicio: Optional[date] = Query(None),
    fecha_fin: Optional[date] = Query(None),
    db: Session = Depends(get_db),
    _=Depends(require_admin),
):
    return shift_service.get_shifts(db, employee_id=employee_id,
                                    fecha_inicio=fecha_inicio, fecha_fin=fecha_fin)


@router.post("", response_model=ShiftOut, status_code=201)
def create_shift(data: ShiftCreate, db: Session = Depends(get_db), _=Depends(require_admin)):
    return shift_service.create_shift(
        db,
        employee_id=data.employee_id,
        fecha=data.fecha,
        hora_inicio=data.hora_inicio,
        hora_fin=data.hora_fin,
    )


@router.get("/{shift_id}", response_model=ShiftOut)
def get_shift(shift_id: int, db: Session = Depends(get_db), _=Depends(require_admin)):
    s = shift_service.get_shift(db, shift_id)
    if not s:
        raise HTTPException(status_code=404, detail="Turno no encontrado")
    return s


@router.put("/{shift_id}", response_model=ShiftOut)
def update_shift(shift_id: int, data: ShiftUpdate,
                 db: Session = Depends(get_db), _=Depends(require_admin)):
    s = shift_service.update_shift(db, shift_id, data.model_dump(exclude_none=True))
    if not s:
        raise HTTPException(status_code=404, detail="Turno no encontrado")
    return s


@router.delete("/{shift_id}", status_code=204)
def delete_shift(shift_id: int, db: Session = Depends(get_db), _=Depends(require_admin)):
    s = shift_service.get_shift(db, shift_id)
    if not s:
        raise HTTPException(status_code=404, detail="Turno no encontrado")
    shift_service.delete_shift(db, shift_id)

File: frontend/src/api/shifts.ts
import { apiClient } from './client'

export interface ShiftOut {
  id: number
  employee_id: number
  fecha: string
  hora_inicio: string
  hora_fin: string
  horas_trabajadas: string
}

export interface ShiftCreate {
  employee_id: number
  fecha: string
  hora_inicio: string
  hora_fin: string
}

export interface ShiftUpdate {
  fecha?: string
  hora_inicio?: string
  hora_fin?: string
}

export const getShifts = (params?: {
  employee_id?: number
  fecha_inicio?: string
  fecha_fin?: string
}): Promise<ShiftOut[]> => {
  const qs = new URLSearchParams()
  if (params?.employee_id != null) qs.set('employee_id', String(params.employee_id))
  if (params?.fecha_inicio) qs.set('fecha_inicio', params.fecha_inicio)
  if (params?.fecha_fin) qs.set('fecha_fin', params.fecha_fin)
  const query = qs.toString() ? `?${qs.toString()}` : ''
  return apiClient.get(`/shifts${query}`)
}

export const createShift = (data: ShiftCreate): Promise<ShiftOut> =>
  apiClient.post('/shifts', data)

export const updateShift = (id: number, data: ShiftUpdate): Promise<ShiftOut> =>
  apiClient.put(`/shifts/${id}`, data)

export const deleteShift = (id: number): Promise<void> =>
  apiClient.delete(`/shifts/${id}`)
